File: lab7/Dictionary.java
// Aaron Chan
// lab7
// Dictionary.java
// 5/30/16

public class Dictionary implements DictionaryInterface{

    // private types and functions ------------------------------------------------

    // NodeObj
    private class Node{
	String key;
	String value;
	Node left;
	Node right;
	// newNode()
	// constructor for private Node type
	Node(String k, String v) {
	    key = k;
	    value = v;
	    left = right = null;
	}
    } 
    private Node root;
    private int numPairs;

    // Dictionary Constructor
    public Dictionary(){
	root = null;
	numPairs = 0;
    } 

    //findKey()
    //returns node with key k in subtee at R
    Node findKey(Node R, String key){
	if(R==null || (key.equals(R.key)))
	    return R;
	if( (key.compareTo(R.key)<0 )) 
	    return findKey(R.left, key);
	else // strcmp(k, R->key)>0
	    return findKey(R.right, key);
    }


    // findParent()
    // returns parent of N in subtree rooted at R, or returns NULL
    Node findParent(Node N, Node R){
	Node P = null;
	if( N!=R ){
	    P = R;
	    while( P.left!=N && P.right!=N ){
		if((N.key.compareTo(P.key)<0))
		    P = P.left;
		else
		    P = P.right;
	    }
	}   
	return P;
    }


    //findLeftmost()
    //returns leftmost node at subtree of R, or returns NULL
    Node findLeftmost(Node R){
	Node L = R;
	if( L!=null ) for( ; L.left!=null; L=L.left);
	return L;
    }

    //printInOrder()
    //prints key/value pairs at subtree of R of increasing order
    String printInOrder(Node R){
	String S = ""; 
	if( R!=null ){
	    S += printInOrder(R.left);
	    S += R.key + " " + R.value + "\n";
	    S +=  printInOrder(R.right);
	}
	return S; 
    }

    //deleteAll()
    //delete all nodes in subtree N
    void deleteAll(Node N){
	if( N!=null ){
	    deleteAll(N.left);
	    deleteAll(N.right);
	} 
    }

    //public functions------
    
    //isEmpty()
    //returns 1 if S is empty, 0 if not empty
    public boolean isEmpty(){
	return (numPairs == 0);
    }

    //size()
    //returns number of key/value pairs
    public int size(){
	return(numPairs);
    }

    //lookup()
    //returns value such that the key/value pair is in D, null it nothing exist
    public String lookup(String key){
	Node N = findKey(root, key);
	return ( N==null ? null : N.value );
    }

    // insert()
    // inserts new key/value pair into D
    public void insert(String key, String value) throws KeyCollisionException{
	if( findKey(root, key)!=null ){
	    throw new KeyCollisionException("Cannot insert duplicate key");
	}
	Node N = new Node(key, value);
	Node B = null;
	Node A = root;
	while( A!=null ){
	    B = A;
	    if( key.compareTo(A.key)<0 ) 
		A = A.left;
	    else A =  A.right;
	}
	if( B==null ){ 
	    root = N;
	}
	else if( key.compareTo(B.key)<0 ){ 
	    B.left = N;
	}
	else B.right = N;
	numPairs++;
    } 

    // delete()
    // deletes pair with the key k
    public void delete(String key) throws KeyNotFoundException{
	Node N = findKey(root, key);
	Node P;
	Node S;
	if( N==null ){
	    throw new KeyNotFoundException("Cannot delete non-existent keys");
	}
	if( N.left==null && N.right==null ){  // case 1:  N has no children
	    if( N==root ){
		root = null;
	    }else{
		P = findParent(N, root);
		if( P.right==N ) P.right = null;
		else P.left = null;
	    }
	}else if( N.right==null ){            // case 2a:  N has a left child only
	    if( N==root ){
		root = N.left;
	    }else{
		P = findParent(N, root);
		if( P.right==N ) P.right = N.left;
		else P.left = N.left;
	    }
	}else if( N.left==null ){            // case 2b: N has a right child only
	    if( N==root ){
		root = N.right;
	    }else{
		P = findParent(N, root);
		if( P.right==N ) P.right = N.right;
		else P.left = N.right;
	    }
	}else{                       // case 3:  N has two children
	    S = findLeftmost(N.right);
	    N.key = S.key;
	    N.value = S.value;
	    P = findParent(S, N);
	    if( P.right==S ) P.right = S.right;
	    else P.left = S.right;
	}
	numPairs--;
    }

    // makeEmpty()
    // resets D to the empty state.
    public void makeEmpty(){
	deleteAll(root);
	root = null;
	numPairs = 0;
    }

    // printDictionary()
    // prints text representation of D to the file pointed to by out
    public String toString(){
	Node N = root;
	return printInOrder(N);
    }   
} 

File: lab5/Dictionary.c
// Aaron Chan - aahchan
// 5/13/16
// Dictionary.c
// lab5

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

  //NodeObj
  typedef struct NodeObj{
    char* key;
    char* value;
    struct NodeObj* next;
  } NodeObj;

//Node
typedef NodeObj* Node;

//newNode()
//constructor of the Node type
Node newNode(char* k, char* v){
  Node N = malloc(sizeof(NodeObj));
  assert(N!=NULL);
  N->key = k;
  N->value = v;
  N->next = NULL;
  return(N);
}

//freeNode()
//destructor for the Node type

void freeNode(Node* pN){
  if(pN!=NULL && *pN!=NULL ){
    free(*pN);
    *pN = NULL;
  }
}

//DictinoaryObj
typedef struct DictionaryObj{
  Node head;
  Node tail;
  int numItems;
} DictionaryObj;

//----------------------------------------------------------------
//newDictionary()
//constructor for Dictionary type
Dictionary newDictionary(void){
  Dictionary S = malloc(sizeof(DictionaryObj));
  assert(S != NULL);
  S->head = NULL;
  S->tail = NULL;
  S->numItems = 0;
  return S;
}

//freeDictionary()
//destructor for the Dictionary type
void freeDictionary(Dictionary* pS){
  if( pS!=NULL && *pS!=NULL){
    if( !isEmpty(*pS) )makeEmpty(*pS);
    free(*pS);
    *pS = NULL;
  }
}
   

//isEmpty()
//returns 1 (true) if S is empty, 0 (false) otherwise
//pre: none
int isEmpty(Dictionary S){
  if( S==NULL ){
    fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");      exit(EXIT_FAILURE);
  }
  if(S->numItems>0){
    return 0;
  }
  return 1;
}

//size()
//returns the number of(key,value) pairs in S
//pre: none
int size(Dictionary S){
  return S->numItems;
}


//lookup()
//returns the value v suc that (k, v) is in S, or returns NULL if no such value exists
//pre: none
char* lookup(Dictionary S, char* k){
  Node N = S->head;
  if( S == NULL ){
    fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary\n");
    exit(EXIT_FAILURE);
  }
  while(N != NULL){
    if(strcmp(N->key,k)== 0)
      return N->value;
    N = N->next;
  }
  return NULL;;
}

//insert()
//inserts new(k,v) pair into S
//pre: lookup(S,K)!= NULL
void insert(Dictionary S, char* k, char* v){
  if(S->numItems == 0){
    S->head = S->tail = newNode(k, v);
  }else{
    Node N;
    N = newNode(k, v);
    S->tail->next = N;
    S->tail = N;
  }
  S->numItems++;
} 

//makeEmpty()
//reset S to the empty state
//pre: none
void makeEmpty(Dictionary S){
  S->numItems = 0;
  freeNode(&S->head);
}

// delete()
// deletes pair with the key k
// pre: lookup(S, k)!=NULL
void delete(Dictionary S, char* k){
  Node N = S->head;
  if( lookup(S,k) == NULL ){
    fprintf(stderr, "error: key not found\n");
    exit(EXIT_FAILURE);
  }
  if(strcmp(N->key,k)==0){
    Node P = S->head;
    Node G = P->next;
    S->head = G;
    freeNode(&P);
  }else{
    while(N !=NULL && N->next != NULL){
      if(strcmp(N->next->key, k)==0){
	Node G = N;
	Node C = N->next;
	G->next = C->next;
	N=G;
	freeNode(&C);
      }
      N = N->next;
    }
  }
  S->numItems--;
}



//printDictionary()
//pre: none
//prints a text representation of S to the file pointed to by out
void printDictionary(FILE* out, Dictionary S){
  Node N;
  if( S==NULL ){
    fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  for(N=S->head; N!=NULL; N=N->next) fprintf(out, "%s %s \n" , N->key, N->value);
  fprintf(out, "\n");
}
